proposals/20210501-plugin-system

```diff
@@ -0,0 +1,262 @@
+# Plugin System
+
+## Summary
+
+This is a proposal to create an infrastructure to extend the functionality of the Falco libraries via plugins.
+
+Plugins will allow users to easily extend the functionality of the libraries and, as a consequence, of Falco and any other tool based on the libraries.
+
+This proposal, in partcular, focuses on two types of plugins: source plugins and extractor plugins. 
+
+## Motivation
+
+libscap and libsinsp provide a powerful data capture framework, with a rich set of features that includes:
+- data capture
+- trace files management
+- enrichment
+- filtering
+- formatting and screen rendering
+- Lua scripting (chisels)
+
+These features have beed designed with one specific input in mind: system calls. However, they are generically adaptable to a broad set of inputs, such as cloud logs.
+
+With this proposal, we want to dramatically extend the scope of what the libraries, Falco and other tools can be applied to. We want to do it in a way that is easy, efficient and empowers anyone in the community to write an extension.
+
+## Goals
+
+- To design and implement a plugin framework that makes the libraries more modular and extensible
+- To have a framework that is easy to use
+- To support dynamic loading of plugins, so that the libraries can be extended without having to be recompiled and relinked
+- To enable users to write plugins in any language, with particular focus on go, C and C++
+- To have an efficient plugin framework so that, performance-wise, writing a plugin is as close as possible as extending the libraries internal source code
+- To make it possible to write plugins for Linux, MacOS and Windows
+
+## Non-Goals
+
+- To implement plugins other than source and extractor: to be approached as separate task
+- To document the plugin framework and interface: to be approached as separate task
+
+## Proposal
+
+### Use cases
+
+- capture events from arbitrary sources and inject them into the libraries for processing --> **source plugins**
+- export and populate filtercheck-compatible fields from the captured events, that can be used in filters, Falco rules, chisels and more --> **extractor plugins**
+- save the captured events into trace files that seamlessly encode all of the context required to parse the data
+
+### Plugins format
+
+Plugins are dynamic libraries (.so files in Unix, .dll files in windows) that reside in specific well known locations and export a minimum set of functions that the libraries will recognize.
```

Comment:
What are these well-known locations?

Reply:
I added a section "Loading plugins in falcosecurity/falco" with more details and example syntax. In short, it's below the "share" directory e.g. /usr/share/falco/plugins